AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy

```diff
@@ -0,0 +1,612 @@
+     Types.MutatedBranchKeyItem(ItemType := "Mutation Index: " + mutationCommitmentUUID, Description := "Created"),
+     Types.MutatedBranchKeyItem(ItemType := "Active: " + newVersion, Description := "Created"),
+     Types.MutatedBranchKeyItem(ItemType := "Decrypt Only: " + newVersion, Description := "Created"),
+     Types.MutatedBranchKeyItem(ItemType := "Beacon", Description := "Mutated")
+    ]
+  }
+
+  method {:isolate_assertions} CreateNewTerminalDecryptOnlyBranchKey(
+    decryptOnlyEncryptionContext: Structure.BranchKeyContext,
+    mutationToApply: StateStrucs.MutationToApply,
+    keyManagerStrategy: KmsUtils.keyManagerStrat
+  )
+    returns (res: Result<KeyStoreTypes.EncryptedHierarchicalKey, Types.Error>)
+    requires KmsArn.ValidKmsArn?(mutationToApply.Terminal.kmsArn)
+    requires KMSKeystoreOperations.AttemptKmsOperation?(
+               KeyStoreTypes.kmsKeyArn(mutationToApply.Terminal.kmsArn), decryptOnlyEncryptionContext
+             )
+    requires keyManagerStrategy.ValidState()
+    modifies
+      match keyManagerStrategy
+      case reEncrypt(kms) => kms.kmsClient.Modifies
+      case decryptEncrypt(kmsD, kmsE) => kmsD.kmsClient.Modifies + kmsE.kmsClient.Modifies
+    ensures keyManagerStrategy.ValidState()
+    ensures res.Success? ==>
+              && Structure.BranchKeyContext?(res.value.EncryptionContext)
+              && Structure.EncryptedHierarchicalKey?(res.value)
+              && res.value.KmsArn == KMSKeystoreOperations.GetKeyId(KeyStoreTypes.kmsKeyArn(mutationToApply.Terminal.kmsArn))
+              && Structure.BRANCH_KEY_TYPE_PREFIX < res.value.EncryptionContext[Structure.TYPE_FIELD]
+              && Structure.BRANCH_KEY_ACTIVE_VERSION_FIELD !in decryptOnlyEncryptionContext
+  {
+    var grantTokens: KMS.GrantTokenList;
+    var kmsClient: KMS.IKMSClient;
+    match keyManagerStrategy {
+      case reEncrypt(kms) =>
+        grantTokens := kms.grantTokens;
+        kmsClient := kms.kmsClient;
+      case decryptEncrypt(kmsD, kmsE) =>
+        grantTokens := kmsE.grantTokens;
+        kmsClient := kmsE.kmsClient;
+    }
+
+    var wrappedDecryptOnlyBranchKey? := KMSKeystoreOperations.GenerateKey(
+      encryptionContext := decryptOnlyEncryptionContext,
+      kmsConfiguration := KeyStoreTypes.kmsKeyArn(mutationToApply.Terminal.kmsArn),
+      grantTokens := grantTokens,
+      kmsClient := kmsClient
+    );
+    var wrappedDecryptOnlyBranchKey :- wrappedDecryptOnlyBranchKey?
+    .MapFailure(e => Types.Error.AwsCryptographyKeyStore(e));
+
+    var newDecryptOnly := Structure.ConstructEncryptedHierarchicalKey(
+      decryptOnlyEncryptionContext,
+      wrappedDecryptOnlyBranchKey.CiphertextBlob.value
+    );
+
+    :- Need(
+      Structure.BRANCH_KEY_TYPE_PREFIX < newDecryptOnly.EncryptionContext[Structure.TYPE_FIELD],
+      Types.KeyStoreAdminException(message := "Invalid Branch Key prefix.")
+    );
```

Comment:
nit: why don't we remove this runtime check and make it an ensures in the `Structure.ConstructEncryptedHierarchicalKey` function?

Reply:
That would limit `ConstructEncryptedHierarchicalKey` to only making versions.
Right now, it can Construct Beacon, Active, and Version.

I think there is another way to get this fix in.
Instead, there could be a requires on the `decryptOnlyEncryptionContext` that it be for a Version.
And then there could be an Ensures clause on the successful output being a Version.

This is what I did for the `NewActiveItemForDecryptEncrypt` method.

```suggestion
    :- Need(
      Structure.BRANCH_KEY_TYPE_PREFIX < newDecryptOnly.EncryptionContext[Structure.TYPE_FIELD],
      Types.KeyStoreAdminException(message := "Invalid Branch Key prefix.")
    );
    // TODO-Mutations-FF : require Decrypt Only Encryption Context
    // TODO-Mutations-FF : ensure Decrypt Only Item
```